dali/pipeline/data/tensor_vector.cc

```diff
@@ -346,57 +368,52 @@ TensorVector<Backend> &TensorVector<Backend>::operator=(TensorVector<Backend> &&
   return *this;
 }
 
+
 template <typename Backend>
 void TensorVector<Backend>::UpdateViews() {
   // Return if we do not have a valid allocation
   if (!IsValidType(tl_->type())) return;
   // we need to be able to share empty view as well so don't check if tl_ has any data
   type_ = tl_->type();
 
-  tensors_.resize(tl_->ntensor());
+  assert(tensors_.size() == tl_->ntensor());
 
   views_count_ = tensors_.size();
   for (size_t i = 0; i < tensors_.size(); i++) {
     update_view(i);
   }
 }
+
+
 template <typename Backend>
 std::shared_ptr<TensorList<Backend>> TensorVector<Backend>::AsTensorList(bool check_contiguity) {
   DALI_ENFORCE(IsContiguous() || !check_contiguity,
-               "Cannot cast non continuous TensorVector "
-               "to TensorList.");
+               "Cannot cast non continuous TensorVector to TensorList.");
   return tl_;
 }
 
-template <typename Backend>
-shared_ptr<Tensor<Backend>> TensorVector<Backend>::create_tensor() const {
-  auto t = std::make_shared<Tensor<Backend>>();
-  t->set_pinned(pinned_);
-  if (IsValidType(type_)) {
-    t->set_type(type_);
-  }
-  return t;
-}
 
 template <typename Backend>
-void TensorVector<Backend>::allocate_tensors(int batch_size) {
-  DALI_ENFORCE(tensors_.empty(), "Changing the batch size is prohibited. It should be set once.");
-  // If we didn't declare the batch size but tried to pin memory or set type
-  // we need to apply it to tensors
-  tensors_.resize(batch_size, nullptr);
-  for (auto &t : tensors_) {
-    t = create_tensor();
+void TensorVector<Backend>::resize_tensors(int new_size) {
+  auto old_size = tensors_.size();
+  tensors_.resize(new_size);
```

Comment:
This is not a vector-like resize. You should either store these tensors in some kind of free list or maintain a separate actual size and never shrink `tensors_`.

Reply:
Added `curr_tensors_size_`